### PR TITLE
Add journalTool support for reading new checkpoint format

### PR DESCRIPTION
**Pull request by aaudiber**
This PR updates the journalTool to be able to read the our new types of journal checkpoints. Instead of writing its output to stdout, the journal tool will now write to a journal dump directory with the below structure:

➜  alluxio git:(35ccca7107) ✗ tree journal_dump-1554762179525
journal_dump-1554762179525
├── checkpoints-377279
│   ├── ACTIVE_SYNC_MANAGER
│   ├── INODE_DIRECTORY_ID_GENERATOR
│   ├── INODE_TREE
│   │   ├── CACHING_INODE_STORE
│   │   │   ├── meta
│   │   │   │   └── 9
│   │   │   ├── private
│   │   │   │   └── 9
│   │   │   │       ├── CURRENT
│   │   │   │       ├── MANIFEST-000004
│   │   │   │       └── OPTIONS-000010
│   │   │   └── shared
│   │   │       ├── 000011.sst
│   │   │       ├── 000012.sst
│   │   │       ├── 000013.sst
│   │   │       ├── 000014.sst
│   │   │       ├── 000015.sst
│   │   │       └── 000016.sst
│   │   ├── INODE_COUNTER
│   │   ├── PINNED_INODE_FILE_IDS
│   │   ├── REPLICATION_LIMITED_FILE_IDS
│   │   ├── TO_BE_PERSISTED_FILE_IDS
│   │   └── TTL_BUCKET_LIST
│   ├── MASTER_UFS_MANAGER
│   └── MOUNT_TABLE
└── edits.txt

To support testing the journal tool with RocksDB checkpoints, this PR also updates our InodeStore/BlockStore code so that AlluxioMaster and SecondaryMaster can exist on the same machine without conflict in their metastore directories.

### CODE REVIEW

File: core/server/common/src/main/java/alluxio/master/journal/checkpoint/CompoundCheckpointFormat.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+  @Override
+  public void parseToHumanReadable(CheckpointInputStream in, PrintStream out) throws IOException {
+    CompoundCheckpointReader reader = createReader(in);
+    Optional<Entry> entryOpt;
+    while ((entryOpt = reader.nextCheckpoint()).isPresent()) {
+      Entry entry = entryOpt.get();
+      out.printf("--- Begin checkpoint for %s ---", entry.getName());
+      CheckpointFormat format = entry.getStream().getType().getCheckpointFormat();
+      format.parseToHumanReadable(entry.getStream(), out);
```
Comment (calvinjia) [inline core/server/common/src/main/java/alluxio/master/journal/checkpoint/CompoundCheckpointFormat.java]: Should we handle the `TarballCheckpointFormat` here?

File: core/server/common/src/main/java/alluxio/master/journal/checkpoint/CheckpointFormat.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  /**
+   * @param in a checkpoint input stream
+   * @param out a print stream for writing a human-readable checkpoint representation
+   */
+  void parseToHumanReadable(CheckpointInputStream in, PrintStream out) throws IOException;
+
+  /**
+   * Interface for checkpoint readers.
+   */
+  interface CheckpointReader {}
```
Comment (calvinjia) [inline core/server/common/src/main/java/alluxio/master/journal/checkpoint/CheckpointFormat.java]: Checkpoint readers are not thread safe?

File: core/server/common/src/main/java/alluxio/master/journal/checkpoint/LongsCheckpointFormat.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  public LongsCheckpointReader createReader(CheckpointInputStream in) {
+    return new LongsCheckpointReader(in);
+  }
+
+  @Override
+  public void parseToHumanReadable(CheckpointInputStream in, PrintStream out) throws IOException {
+    LongsCheckpointReader reader = createReader(in);
+    Optional<Long> longOpt;
+    while ((longOpt = reader.nextLong()).isPresent()) {
+      out.printf("%d ", longOpt.get());
```
Comment (calvinjia) [inline core/server/common/src/main/java/alluxio/master/journal/checkpoint/LongsCheckpointFormat.java]: New line vs space delimited, new line might be better from opening the file perspective?

File: core/server/common/src/main/java/alluxio/master/journal/checkpoint/CheckpointFormat.java (hunk 2 above)
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/journal/checkpoint/CheckpointFormat.java]: This interface has no methods?

File: core/server/common/src/main/java/alluxio/master/journal/checkpoint/CompoundCheckpointFormat.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    return new CompoundCheckpointReader(in);
+  }
+
+  @Override
+  public void parseToHumanReadable(CheckpointInputStream in, PrintStream out) throws IOException {
+    CompoundCheckpointReader reader = createReader(in);
+    Optional<Entry> entryOpt;
+    while ((entryOpt = reader.nextCheckpoint()).isPresent()) {
+      Entry entry = entryOpt.get();
+      out.printf("--- Begin checkpoint for %s ---", entry.getName());
```
Comment (gpang) [inline core/server/common/src/main/java/alluxio/master/journal/checkpoint/CompoundCheckpointFormat.java]: should these have newlines?

File: core/server/common/src/main/java/alluxio/master/journal/checkpoint/CompoundCheckpointFormat.java (hunk 1 above)
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/checkpoint/CompoundCheckpointFormat.java]: it handles it by calling `TarballCheckpointFormat#parseToHumanReadable`, which tells the user to use the journal tool for more detail. There isn't a great way to represent a tarball in a `PrintStream`.

File: core/server/common/src/main/java/alluxio/master/journal/checkpoint/CheckpointFormat.java (hunk 2 above)
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/checkpoint/CheckpointFormat.java]: Yep, added an annotation

File: core/server/common/src/main/java/alluxio/master/journal/checkpoint/LongsCheckpointFormat.java (hunk 3 above)
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/checkpoint/LongsCheckpointFormat.java]: good call, switched to newline

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 5]
```diff
@@ -1764,6 +1764,18 @@ public String toString() {
...
+  //
+  // Secondary master related properties
+  //
+  public static final PropertyKey SECONDARY_MASTER_METASTORE_DIR =
+      new Builder(Name.SECONDARY_MASTER_METASTORE_DIR)
+          .setDefaultValue(String.format("${%s}/secondary-metastore", Name.WORK_DIR))
+          .setDescription(
+              "The secondary master metastore work directory. Only some metastores need disk.")
+          .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
+          .setScope(Scope.NONE)
```
[the remark this answers is not in the document]
Comment (aaudiber) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: It affects only `SecondaryMaster`, which is a separate process type from `AlluxioMaster`. I followed the example of our `Proxy` property keys, where we set the scope to `NONE`

File: core/server/common/src/main/java/alluxio/master/journal/JournalTool.java [hunk 6]
```diff
@@ -98,69 +104,32 @@ public static void main(String[] args) {
...
-  }
-
-  private static void readFromJournal() {
+  private static void dumpJournal() throws IOException {
+    System.out.printf("Dumping journal to %s%n", sOutputDir);
+    Files.createDirectories(Paths.get(sOutputDir));
     UfsJournal journal =
         new UfsJournalSystem(getJournalLocation(), 0).createJournal(new NoopMaster(sMaster));
-    try (JournalReader reader = new UfsJournalReader(journal, sStart, true)) {
+    try (PrintStream entryStream =
```
[the remark this answers is not in the document]
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/JournalTool.java]: changed to `out`

File: core/server/common/src/main/java/alluxio/master/journal/checkpoint/CheckpointFormat.java (hunk 2 above)
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/checkpoint/CheckpointFormat.java]: It's a marker interface for checkpoint readers. We could have `createReader` return `Object`, instead, but I think this way expresses the intention better.

File: core/server/common/src/main/java/alluxio/master/journal/checkpoint/CompoundCheckpointFormat.java (hunk 4 above)
Comment (aaudiber) [reply core/server/common/src/main/java/alluxio/master/journal/checkpoint/CompoundCheckpointFormat.java]: good catch, fixed

**Comment by aaudiber** [comment]
@calvinjia @gpang PTAL

**Comment by aaudiber** [comment]
alluxio-bot, merge this please

**Pull request merged**